stateupgrade/state_upgrade.go

```diff
@@ -0,0 +1,46 @@
+// (c) 2023 Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package stateupgrade
+
+import (
+	"math/big"
+
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Configure applies the state upgrade to the state.
+func Configure(stateUpgrade *params.StateUpgrade, chainConfig ChainContext, state StateDB, blockContext BlockContext) error {
+	isEIP158 := chainConfig.IsEIP158(blockContext.Number())
+	for account, upgrade := range stateUpgrade.StateUpgradeAccounts {
+		if err := upgradeAccount(account, upgrade, state, isEIP158); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// upgradeAccount applies the state upgrade to the given account.
+func upgradeAccount(account common.Address, upgrade params.StateUpgradeAccount, state StateDB, isEIP158 bool) error {
+	// Create the account if it does not exist
+	if !state.Exist(account) {
+		state.CreateAccount(account)
+	}
+
+	if upgrade.BalanceChange != nil {
+		state.AddBalance(account, (*big.Int)(upgrade.BalanceChange))
+	}
+	if len(upgrade.Code) != 0 {
+		// if the nonce is 0, set the nonce to 1 as we would when deploying a contract at
+		// the address.
+		if isEIP158 && state.GetNonce(account) == 0 {
+			state.SetNonce(account, 1)
```

Comment:
are we setting nonce = 1 to prevent account deletion according to EIP158?

Reply:
If we set the code, why do we need lines 37-38?

Reply:
Let me know if I am misunderstanding something. https://github.com/ethereum/EIPs/blob/master/EIPS/eip-158.md

Reply:
this code was added for symmetry with https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go?rgh-link-date=2023-03-01T10%3A25%3A28Z#L526

Reply:
Right [that](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go?rgh-link-date=2023-03-01T10%3A25%3A28Z#L526) makes sense to increase the nonce after creation since it is considered ["empty".](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-158.md) However, we set the account code to some codebytes which makes it not "empty". Not sure why we are also increasing nonce as well

Reply:
true. at the end, a contract deployed to a new address with create will have nonce = 1 and if we deploy a new contract via the state upgrade I think it is better to also have nonce = 1